src/dns-test.js

```diff
@@ -0,0 +1,317 @@
+}
+
+/**
+ * Sleep implementation
+ */
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+/**
+ * Read the client's nameservers from disk.
+ * If on macOS, read /etc/resolv.comf.
+ * If on Windows, read a registry.
+ */
+async function readNameservers() {
+    let nameservers = [];
+    try { 
+        let platform = await browser.runtime.getPlatformInfo();
+        if (platform.os == "mac") {
+            nameservers = await browser.experiments.resolvconf.readNameserversMac();
+        } else if (platform.os == "win") {
+            nameservers = await browser.experiments.resolvconf.readNameserversWin();
+        } else {
+            sendTelemetry({"reason": "osNotSupportedError"});
+            throw new Error("DNSSEC Interference Study: OS not supported");
+        }
+    } catch(e) {
+        sendTelemetry({"reason": "readNameserversFileError"});
+        throw new Error("DNSSEC Interference Study: Couldn't find nameservers file");
+    } 
+
+    if (!Array.isArray(nameservers) || nameservers.length <= 0) {
+        sendTelemetry({"reason": "noNameserversInFileError"});
+        throw new Error("No nameservers found in /etc/resolv.conf or registry");
+    }
+
+    let nameservers_ipv4 = [];
+    for (var i = 0; i < nameservers.length; i++) {
+        let ns = nameservers[i];
+        if (!isUndefined(ns) && ns.includes(".")) {
+            nameservers_ipv4.push(ns);
+        }
+    }
+
+    if (nameservers_ipv4.length <= 0) {
+        sendTelemetry({"reason": "noIPv4NameserversError"});
+        throw new Error("DNSSEC Interference Study: No IPv4 nameservers found");
+    }
+    return nameservers_ipv4;
+}
+
+/**
+ * Open the UDP sockets and add an event listener for when we receive UDP 
+ * responses.
+ */
+async function setupUDPCode() {
+    try {
+        await browser.experiments.udpsocket.openSocket();
+        browser.experiments.udpsocket.onDNSResponseReceived.addListener(processUDPResponse);
+    } catch(e) {
+        sendTelemetry({"reason": "openUDPSocketsError"});
+        throw new Error("DNSSEC Interference Study: Couldn't set up UDP socket or reason listener");
+    }
+}
+
+/**
+ * For each RR type that we have a DNS record for, attempt to send queries over 
+ * UDP and TCP.
+ */
+async function sendQueries(nameservers_ipv4) {
+    for (let i = 0; i < RRTYPES.length; i++) {
+        let rrtype = RRTYPES[i];
+        if (rrtype == 'SMIMEA') {
+            await sendUDPQuery(SMIMEA_DOMAIN_NAME, nameservers_ipv4, rrtype);
+            await sendTCPQuery(SMIMEA_DOMAIN_NAME, nameservers_ipv4, rrtype);
+        } else if (rrtype == 'HTTPS') {
+            await sendUDPQuery(HTTPS_DOMAIN_NAME, nameservers_ipv4, rrtype);
+            await sendTCPQuery(HTTPS_DOMAIN_NAME, nameservers_ipv4, rrtype);
+        } else {
+            await sendUDPQuery(APEX_DOMAIN_NAME, nameservers_ipv4, rrtype);
+            await sendTCPQuery(APEX_DOMAIN_NAME, nameservers_ipv4, rrtype);
+        }
+    }
```

Comment:
Do you intentionally send one request at a time, or would you prefer to run them in parallel if possible?

If the latter, then you can use `RRTYPES.map(async rrtype => { ... })` to create a list of promises and then `await Promise.all(...)` to wait for all promises in the list to settle.

Reply:
Right now, I intentionally send one request at a time because it simplifies the design. I suppose I could run them in parallel with the current design of the `experiment.udpsocket` and `experiment.tcpsocket` APIs. However, if I were to re-design the experiments.udpsocket API to use a single socket, it's not clear to me how I could send multiple queries in parallel over the socket and associate each response (or lack thereof) out-of-order with an RRTYPE.

Reply:
> However, if I were to re-design the experiments.udpsocket API to use a single socket, then it's not clear to me how I could send multiple queries in parallel over the socket

I think that you have misunderstood my suggestion at https://github.com/mozilla-extensions/dnssec-interference/pull/5#discussion_r468146588 . I'm not recommending to use one socket shared by every API call, but rather to create a new socket instance - to have an API that completely abstracts away the fact that sockets are used internally. Your current tcpsocket implementation already uses one socket, the same can be done with udpsocket.

Reply:
See my comment about simplifying the udpsocket API [here](https://github.com/mozilla-extensions/dnssec-interference/pull/5#discussion_r469547340).

Reply:
I think I'm going to send one request at a time for the sake of simplicity for now, but I may look at this later.